File: api/src/main/java/mb/pie/api/stamp/resource/Hash.java
package mb.pie.api.stamp.resource;

import mb.resource.ReadableResource;
import mb.resource.hierarchical.HierarchicalResource;
import mb.resource.hierarchical.match.ResourceMatcher;
import mb.resource.hierarchical.walk.ResourceWalker;
import org.checkerframework.checker.nullness.qual.Nullable;

import java.io.IOException;
import java.io.UncheckedIOException;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.stream.Stream;

public class Hash {
    private final MessageDigest digest;


    Hash() {
        try {
            digest = MessageDigest.getInstance("SHA-1");
        } catch(NoSuchAlgorithmException e) {
            throw new RuntimeException(e);
        }
    }


    void update(ReadableResource resource) throws IOException {
        if(!resource.exists()) {
            updateNonExistent();
        } else {
            updateResource(resource);
        }
    }

    void update(HierarchicalResource resource, @Nullable ResourceMatcher matcher) throws IOException {
        if(!resource.exists()) {
            updateNonExistent();
        } else if(resource.isFile()) {
            updateResource(resource);
        } else if(resource.isDirectory()) {
            updateDir(resource, matcher);
        }
    }

    void updateRec(HierarchicalResource resource, @Nullable ResourceWalker walker, @Nullable ResourceMatcher matcher) throws IOException {
        if(!resource.exists()) {
            updateNonExistent();
        } else if(resource.isFile()) {
            updateResource(resource);
        } else if(resource.isDirectory()) {
            updateDirRec(resource, walker, matcher);
        }
    }


    private void updateNonExistent() {
        digest.update((byte) 0);
    }

    private void updateResource(ReadableResource file) throws IOException {
        digest.update((byte) 1);
        digest.update(file.readBytes());
    }

    private void updateDir(HierarchicalResource dir, @Nullable ResourceMatcher matcher) throws IOException {
        final boolean useMatcher = matcher != null;
        try(final Stream<? extends HierarchicalResource> stream = useMatcher ? dir.list(matcher) : dir.list()) {
            stream.forEach((resource) -> {
                try {
                    updateResource(resource);
                } catch(IOException e) {
                    throw new UncheckedIOException(e);
                }
            });
        } catch(UncheckedIOException e) {
            throw e.getCause();
        }
    }

    private void updateDirRec(HierarchicalResource dir, @Nullable ResourceWalker walker, @Nullable ResourceMatcher matcher) throws IOException {
        final boolean useWalker = walker != null && matcher != null;
        try(final Stream<? extends HierarchicalResource> stream = useWalker ? dir.walk(walker, matcher) : dir.walk()) {
            stream.forEach((resource) -> {
                try {
                    updateResource(resource);
                } catch(IOException e) {
                    throw new UncheckedIOException(e);
                }
            });
        } catch(UncheckedIOException e) {
            throw e.getCause();
        }
    }

    byte[] getHashBytesAndReset() {
        return digest.digest();
    }
}

File: api/src/main/kotlin/mb/pie/api/fs/stamp/Hash.kt
package mb.pie.api.fs.stamp

import mb.fs.api.node.*
import mb.fs.api.node.match.FSNodeMatcher
import mb.fs.api.node.walk.FSNodeWalker
import mb.pie.api.fs.FileSystemResource
import java.security.MessageDigest

interface HashResourceStamperTrait : FileSystemStamper {
  fun createDigester() = MessageDigest.getInstance("SHA-1")!!

  fun MessageDigest.update(node: FSNode, matcher: FSNodeMatcher?) {
    if(node.isDirectory) updateDir(node, matcher)
    if(node.isFile) updateFile(node)
  }

  fun MessageDigest.updateRec(node: FSNode, walker: FSNodeWalker?, matcher: FSNodeMatcher?) {
    if(node.isDirectory) updateDirRec(node, walker, matcher)
    if(node.isFile) updateFile(node)
  }

  fun MessageDigest.updateDir(dir: FSNode, matcher: FSNodeMatcher?) {
    if(matcher == null) {
      dir.list()
    } else {
      dir.list(matcher)
    }.use { stream ->
      for(node in stream) {
        if(node.isFile) updateFile(node)
      }
    }
  }

  fun MessageDigest.updateDirRec(dir: FSNode, walker: FSNodeWalker?, matcher: FSNodeMatcher?) {
    if(walker == null || matcher == null) {
      dir.walk()
    } else {
      dir.walk(walker, matcher)
    }.use { stream ->
      for(node in stream) {
        if(node.isFile) updateFile(node)
      }
    }
  }

  fun MessageDigest.updateFile(file: FSNode) {
    update(file.readAllBytes())
  }
}

data class HashResourceStamper @JvmOverloads constructor(
  private val matcher: FSNodeMatcher? = null
) : HashResourceStamperTrait {
  override fun stamp(resource: FileSystemResource): FileSystemStamp {
    val node = resource.node
    if(!node.exists()) {
      return ByteArrayResourceStamp(null, this)
    }
    val digest = createDigester()
    digest.update(node, matcher)
    val bytes = digest.digest()
    return ByteArrayResourceStamp(bytes, this)
  }

  override fun toString(): String {
    return "Hash($matcher)"
  }
}

data class RecHashResourceStamper @JvmOverloads constructor(
  private val walker: FSNodeWalker? = null,
  private val matcher: FSNodeMatcher? = null
) : HashResourceStamperTrait {
  override fun stamp(resource: FileSystemResource): FileSystemStamp {
    val node = resource.node
    if(!node.exists()) {
      return ByteArrayResourceStamp(null, this)
    }
    val digest = createDigester()
    digest.updateRec(node, walker, matcher)
    val bytes = digest.digest()
    return ByteArrayResourceStamp(bytes, this)
  }

  override fun toString(): String {
    return "RecHash($walker)"
  }
}